plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/plugins/ToolkitUpdateManager.kt

```diff
@@ -0,0 +1,162 @@
+import software.aws.toolkits.jetbrains.utils.notifyInfo
+import software.aws.toolkits.resources.message
+import software.aws.toolkits.telemetry.Component
+import software.aws.toolkits.telemetry.Result
+import software.aws.toolkits.telemetry.ToolkitTelemetry
+
+class ToolkitUpdateManager {
+    private val alarm = Alarm(Alarm.ThreadToUse.SWING_THREAD)
+
+    init {
+        runTask()
+    }
+
+    @VisibleForTesting
+    internal fun runTask() {
+        if (alarm.isDisposed) return
+        scheduleUpdateTask()
+
+        val enabled = AwsSettings.getInstance().isAutoUpdateEnabled
+        LOG.debug { "AWS Toolkit checking for new updates. Auto update enabled: $enabled" }
+
+        if (!enabled) return
+
+        runInEdt {
+            ProgressManager.getInstance().run(
+                object : Task.Backgroundable(null, message("aws.settings.auto_update.progress.message")) {
+                    override fun run(indicator: ProgressIndicator) {
+                        checkForUpdates(indicator)
+                    }
+                }
+            )
+        }
+    }
+
+    private fun scheduleUpdateTask() {
+        alarm.addRequest({ runTask() }, UPDATE_CHECK_INTERVAL_IN_MS)
+    }
+
+    @RequiresBackgroundThread
+    fun checkForUpdates(progressIndicator: ProgressIndicator) {
+        // Note: This will need to handle exceptions and ensure thread-safety
+        try {
+            val toolkitPlugin = AwsToolkit.DESCRIPTOR as IdeaPluginDescriptor? ?: return
+            if (!toolkitPlugin.isEnabled) {
+                LOG.debug { "AWS Toolkit is disabled, not performing auto-update" }
+                return
+            }
+            LOG.debug { "Current version: ${toolkitPlugin.version}" }
+            val latestToolkitPluginDownloader = getUpdate(toolkitPlugin)
+            if (latestToolkitPluginDownloader == null) {
+                LOG.debug { "No newer version found, not performing auto-update" }
+                return
+            } else {
+                LOG.debug { "Found newer version: ${latestToolkitPluginDownloader.pluginVersion}" }
+            }
+
+            if (!latestToolkitPluginDownloader.prepareToInstall(progressIndicator)) return
+            latestToolkitPluginDownloader.install()
+            ToolkitTelemetry.showAction(
+                project = null,
+                success = true,
+                id = SOURCE_AUTO_UPDATE_FINISH_NOTIFY,
+                source = SOURCE_AUTO_UPDATE_FINISH_NOTIFY,
+                component = Component.Filesystem
+            )
+        } catch (e: Exception) {
+            LOG.debug(e) { "Unable to install AWS Toolkit" }
+            ToolkitTelemetry.showAction(
+                project = null,
+                success = false,
+                id = SOURCE_AUTO_UPDATE_FINISH_NOTIFY,
+                source = SOURCE_AUTO_UPDATE_FINISH_NOTIFY,
+                component = Component.Filesystem,
+                reason = e.message
```

Comment:
would like to know if this reason has a length validation check that `e.message` could potentially violate

Reply:
i think you need to keep it <1k
@dhruvigajjar might have some input

Reply:
Sync'ed offline that requirements are <64k